src/ChaiVM/interpreter/code-manager/code-manager.cpp

```diff
@@ -0,0 +1,141 @@
+#include <bit>
+#include <cassert>
+
+#include "ChaiVM/interpreter/code-manager/code-manager.hpp"
+
+namespace chai::interpreter {
+
+void CodeManager::load(const std::filesystem::path &path) {
+    std::ifstream input_file(path, std::ios::binary | std::ios::in);
+    if (input_file.good() && input_file.is_open()) {
+        loadPool(input_file);
+        Immidiate func_count = 0;
+        input_file.read(reinterpret_cast<char *>(&func_count),
+                        sizeof func_count);
+        funcs_ = std::vector<Function>{};
+        for (int i = 0; i < func_count; ++i) {
+            loadFunction(input_file);
+        }
+        input_file.close();
+    } else {
+        throw std::invalid_argument(std::string{"Invalid path "} +
+                                    path.string());
+    }
+}
+
+void CodeManager::loadPool(std::istream &istream) {
+    if (!istream.good()) {
+        throw std::invalid_argument(std::string{"Bad input stream"});
+    }
+    Immidiate constants = 0;
+    istream.read(reinterpret_cast<char *>(&constants), sizeof constants);
+    dispatch_ = std::vector<Immidiate>(constants, -1);
+    for (int i = 0; i < constants; ++i) {
+        char type;
+        istream.read(&type, sizeof type);
+        switch (type) {
+        case CNST_I64:
+            int64_t next_long;
+            istream.read(reinterpret_cast<char *>(&next_long),
+                         sizeof next_long);
+            constantPool_.push_back(next_long);
+            break;
+        case CNST_F64:
+            double next_d;
+            istream.read(reinterpret_cast<char *>(&next_d), sizeof next_d);
+            constantPool_.push_back(std::bit_cast<chsize_t>(next_d));
+            break;
+        case CNST_FUNC_NAME_AND_TYPE:
+            Immidiate name_index;
+            istream.read(reinterpret_cast<char *>(&name_index),
+                         sizeof name_index);
+            Immidiate descriptor_index;
+            istream.read(reinterpret_cast<char *>(&descriptor_index),
+                         sizeof descriptor_index);
+            constantPool_.push_back((static_cast<chsize_t>(type) << 32) |
+                                    (static_cast<chsize_t>(name_index) << 16) |
+                                    (static_cast<chsize_t>(name_index) << 0));
+            break;
+        case CNST_RAW_STR:
+            uint16_t len;
+            istream.read(reinterpret_cast<char *>(&len), sizeof len);
+            char *buf;
+            buf = new char[len + 1];
+            istream.read(buf, len);
+            buf[len] = 0;
```

Comment:
@c71n93 this code is safer since always null-terminate it. If the string will not be null-terminated, the function will not behave unpredictably. Also we avoid redundant contract with assemler here.